yjit/src/codegen.rs

```diff
@@ -3449,6 +3464,26 @@ fn jit_guard_known_klass(
     true
 }
 
+// Guard that the two supplied strings have the same encoding index (see CRuby's encindex.h)
+fn jit_guard_operands_same_encoding(
+    cb: &mut CodeBlock,
+    ocb: &mut OutlinedCb,
+    side_exit: CodePtr,
+    str1_opnd: X86Opnd,  // string1 VALUE
+    str2_opnd: X86Opnd,  // string2 VALUE
+)
+{
+    // Extract str1's string encoding into REG0 and str2's encoding into REG1
+    jit_get_object_flags(cb, REG0, str1_opnd, RUBY_ENCODING_MASK as u64, RUBY_ENCODING_SHIFT as u64);
+    jit_get_object_flags(cb, REG1, str2_opnd, RUBY_ENCODING_MASK as u64, RUBY_ENCODING_SHIFT as u64);
```

Comment:
(non-blocking) We could combine some of these operations (the shift seems particularly redundant), instead of 2x`mov/mov/and/shr` + `cmp` we could (pseudocode):
```
mov REG0, str1_opnd
mov REG1, str2_opnd
mov REG0, mem_opnd(REG0, basic_flags) # get str1's flags
xor REG0, mem_opnd(REG1, basic_flags) # xor with str2's flags
test REG0, RUBY_ENCODING_MASK         # Check if any encoding bits differed

jne ...
```

but please consider this non-blocking as we could address it later, or hopefully in the future we can have an optimizing pass do this for us 🙂

Reply:
That definitely looks more efficient than my initial pass :-)

Reply:
I've used this as the basis for a different implementation.